indico/modules/events/papers/client/js/components/PaperReviewForm.jsx

```diff
@@ -5,114 +5,136 @@
 
 import UserAvatar from 'indico/modules/events/reviewing/components/UserAvatar';
-import {FinalDropdown, FinalSubmitButton, FinalTextArea} from 'indico/react/forms';
 import {Translate} from 'indico/react/i18n';
 
-import {addComment} from '../actions';
-import {canCommentPaper, getPaperDetails, getCurrentUser} from '../selectors';
+import CommentForm from './CommentForm';
+import GroupReviewForm from './GroupReviewForm';
+import {createComment as createCommentAction, createReview, updateReview} from '../actions';
+import {canCommentPaper, canReviewPaper, getPaperDetails, getCurrentUser} from '../selectors';
 
 import './PaperReviewForm.module.scss';
 
-const visibilityOptions = [
-  {
-    value: 'judges',
-    text: Translate.string('Visible only to judges'),
-  },
-  {
-    value: 'reviewers',
-    text: Translate.string('Visible to reviewers and judges'),
-  },
-  {
-    value: 'contributors',
-    text: Translate.string('Visible to contributors, reviewers and judges'),
-  },
-];
-
 export default function PaperReviewForm() {
   const {
     event: {id: eventId},
     contribution: {id: contributionId},
+    lastRevision,
+    isInFinalState,
   } = useSelector(getPaperDetails);
+  const {reviewerData} = lastRevision;
+  const [commentFormExpanded, setCommentFormExpanded] = useState(false);
+  const [reviewGroup, setReviewGroup] = useState(null);
   const user = useSelector(getCurrentUser);
   const canComment = useSelector(canCommentPaper);
+  const canReview = useSelector(canReviewPaper);
   const dispatch = useDispatch();
-  const [commentFormVisible, setCommentFormVisible] = useState(false);
 
-  const onCommentClickHandler = () => {
-    if (!commentFormVisible) {
-      setCommentFormVisible(true);
-    }
-  };
+  const reviewGroups = reviewerData.groups.map(group => group.name);
+  const reviewedGroups = reviewerData.reviewedGroups.map(group => group.name);
 
   const createComment = useCallback(
-    async (formData, form) => {
-      const rv = await dispatch(addComment(eventId, contributionId, formData));
+    async formData => {
+      const rv = await dispatch(createCommentAction(eventId, contributionId, formData));
       if (rv.error) {
         return rv.error;
       }
-      setCommentFormVisible(false);
-      setTimeout(() => form.reset(), 0);
     },
     [dispatch, eventId, contributionId]
   );
 
+  const renderReviewTrigger = () => {
+    return (
+      <div className="review-trigger flexrow">
+        <span className="comment-or-review">
+          <Translate>or</Translate>
+        </span>
+        {reviewGroups.length === 1 ? (
+          <Button onClick={() => setReviewGroup(reviewGroups[0])}>
+            {reviewGroups[0] in reviewerData.reviews ? (
+              <Translate>Change review</Translate>
+            ) : (
+              <Translate>Review</Translate>
+            )}
+          </Button>
+        ) : (
+          <Dropdown
+            text={
+              _.isEqual(reviewGroups, reviewedGroups)
```

Comment:
do they always have the correct order or could this check sometimes fail due to a different order in (one of) the arrays? if the order isn't guaranteed, our `unsortedArraysEqual` util would help

Reply:
I am pretty sure the order is guaranteed but I think it won't hurt to use `unsortedArraysEqual`